File: Sharp.Core/GameEntities/PlayerController.cs
/* 
 * ModSharp
 * Copyright (C) 2023-2025 Kxnrl. All Rights Reserved.
 *
 * This file is part of ModSharp.
 * ModSharp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ModSharp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ModSharp. If not, see <https://www.gnu.org/licenses/>.
 */

using Sharp.Core.Bridges.Natives;
using Sharp.Core.GameObjects;
using Sharp.Core.Helpers;
using Sharp.Core.Utilities;
using Sharp.Generator;
using Sharp.Shared;
using Sharp.Shared.Enums;
using Sharp.Shared.GameEntities;
using Sharp.Shared.GameObjects;
using Sharp.Shared.Types;
using Sharp.Shared.Types.Tier;
using Sharp.Shared.Units;

namespace Sharp.Core.GameEntities;

internal partial class PlayerController : BaseEntity, IPlayerController
{
    public void Print(HudPrintChannel channel,
        string                        message,
        string?                       param1 = null,
        string?                       param2 = null,
        string?                       param3 = null,
        string?                       param4 = null)
        => Player.ControllerPrint(_this,
                                  channel,
                                  channel is HudPrintChannel.Chat && message[0] != '#' ? $" {message}" : message,
                                  param1,
                                  param2,
                                  param3,
                                  param4);

    public float LaggedMovement
    {
        get => Player.ControllerGetLaggedMovement(_this);
        set => Player.ControllerSetLaggedMovement(_this, value);
    }

    public void SwitchTeam(CStrikeTeam team)
        => Player.ControllerSwitchTeam(_this, team);

    public void Respawn()
        => Player.ControllerRoundRespawn(_this);

    public void SetClanTag(string tag)
        => Player.ControllerSetClanTag(_this, tag);

    public void CheckPawn()
        => Player.ControllerCheckPawn(_this);

    public void SetPawn(IPlayerPawn pawn, bool unknown1, bool unknown2, bool unknown3, bool unknown4)
        => Player.ControllerSetPawn(_this, pawn.GetAbsPtr(), unknown1, unknown2, unknown3, unknown4);

    public IPlayerPawn? GetPawn()
        => PlayerPawn.Create(Player.ControllerGetPawn(_this));

    public IPlayerPawn? GetPlayerPawn()
        => PlayerPawn.Create(PlayerPawnHandle.GetEntityPtr());

    public IPlayerPawn? GetObserverPawn()
        => PlayerPawn.Create(ObserverPawnHandle.GetEntityPtr());

    public void SetPlayerPawn(IPlayerPawn pawn)
        => PlayerPawnHandle = pawn.Handle.As<IPlayerPawn>();

    public PlayerSlot PlayerSlot => (byte) (Index - 1);

    public void RefreshScoreBoardData()
    {
        UpdaterCount++;
    }

    public unsafe SoundOpEventGuid EmitSoundClient(string sound, float? volume = null)
        => Player.ControllerEmitSoundClient(_this, sound, &volume);

    public IEconItemView? GetItemInLoadoutFromInventory(CStrikeTeam team, int slot)
        => EconItemView.Create(Player.ControllerGetItemInLoadoutFromInventory(_this, team, slot));

    public string PlayerName
    {
        get => SchemaSystem.GetNetVarString(_this, "CBasePlayerController", "m_iszPlayerName");
        set => SchemaSystem.SetNetVarString(_this, "CBasePlayerController", "m_iszPlayerName", value, 128);
    }

#region Schemas

    [NativeSchemaField("CCSPlayerController", "m_szClan", typeof(CUtlSymbolLarge))]
    public partial SchemaField GetClanTagField();

    [NativeSchemaField("CCSPlayerController", "m_hPlayerPawn", typeof(CEntityHandle<IPlayerPawn>))]
    public partial SchemaField GetPlayerPawnHandleField();

    [NativeSchemaField("CCSPlayerController", "m_hObserverPawn", typeof(CEntityHandle<IPlayerPawn>))]
    public partial SchemaField GetObserverPawnHandleField();

    [NativeSchemaField("CCSPlayerController", "m_iPawnHealth", typeof(uint))]
    public partial SchemaField GetPawnHealthField();

    [NativeSchemaField("CBasePlayerController", "m_bIsHLTV", typeof(bool))]
    public partial SchemaField GetIsHltvField();

    [NativeSchemaField("CCSPlayerController", "m_iPendingTeamNum", typeof(CStrikeTeam))]
    public partial SchemaField GetPendingTeamNumField();

    [NativeSchemaField("CCSPlayerController", "m_bHasCommunicationAbuseMute", typeof(bool))]
    public partial SchemaField GetHasCommunicationAbuseMuteField();

    [NativeSchemaField("CCSPlayerController", "m_bControllingBot", typeof(bool))]
    public partial SchemaField GetControllingBotField();

    [NativeSchemaField("CCSPlayerController", "m_bPawnIsAlive", typeof(bool))]
    public partial SchemaField GetIsPawnAliveField();

    [NativeSchemaField("CBasePlayerController", "m_steamID", typeof(SteamID))]
    public partial SchemaField GetSteamIdField();

    [NativeSchemaField("CCSPlayerController", "m_iScore", typeof(int))]
    public partial SchemaField GetScoreField();

    [NativeSchemaField("CCSPlayerController", "m_iRoundsWon", typeof(int))]
    public partial SchemaField GetRoundsWonField();

    [NativeSchemaField("CCSPlayerController", "m_iMVPs", typeof(int))]
    public partial SchemaField GetMvpCountField();

    [NativeSchemaField("CCSPlayerController", "m_nUpdateCounter", typeof(int))]
    public partial SchemaField GetUpdaterCountField();

    [NativeSchemaField("CCSPlayerController", "m_bSwitchTeamsOnNextRoundReset", typeof(bool))]
    private partial SchemaField GetSwitchTeamsOnNextRoundResetField();

    [NativeSchemaField("CCSPlayerController", "m_bRemoveAllItemsOnNextRoundReset", typeof(bool))]
    private partial SchemaField GetRemoveAllItemsOnNextRoundResetField();

    [NativeSchemaField("CCSPlayerController", "m_iCompetitiveRanking", typeof(int))]
    private partial SchemaField GetCompetitiveRankingField();

    [NativeSchemaField("CCSPlayerController", "m_iCompetitiveWins", typeof(int))]
    private partial SchemaField GetCompetitiveWinsField();

    [NativeSchemaField("CCSPlayerController", "m_iCompetitiveRankType", typeof(CompetitiveRankType))]
    private partial SchemaField GetCompetitiveRankTypeField();

    [NativeSchemaField("CCSPlayerController", "m_DesiredObserverMode", typeof(int))]
    private partial SchemaField GetDesiredObserverModeField();

    [NativeSchemaField("CBasePlayerController", "m_iConnected", typeof(PlayerConnectedState))]
    private partial SchemaField GetConnectedStateField();

    [NativeSchemaField("CCSPlayerController", "m_pDamageServices", typeof(DamageService))]
    private partial SchemaField GetDamageServiceField();

    [NativeSchemaField("CCSPlayerController", "m_pInGameMoneyServices", typeof(InGameMoneyService))]
    private partial SchemaField GetInGameMoneyServiceField();

    [NativeSchemaField("CCSPlayerController", "m_pInventoryServices", typeof(InventoryService))]
    private partial SchemaField GetInventoryServiceField();

    [NativeSchemaField("CCSPlayerController", "m_pActionTrackingServices", typeof(ControllerActionTrackingService))]
    private partial SchemaField GetActionTrackingServiceField();

    [NativeSchemaField("CBasePlayerController", "m_iDesiredFOV", typeof(uint))]
    private partial SchemaField GetDesiredFOVField();

#endregion
}

File: Engine/src/cstrike/entity/PlayerController.cpp
/* 
 * ModSharp
 * Copyright (C) 2023-2025 Kxnrl. All Rights Reserved.
 *
 * This file is part of ModSharp.
 * ModSharp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ModSharp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ModSharp. If not, see <https://www.gnu.org/licenses/>.
 */

#include "gamedata.h"
#include "global.h"
#include "netmessage.h"
#include "vhook/call.h"

#include "cstrike/component/PlayerControllerComponent.h"
#include "cstrike/entity/PlayerController.h"
#include "cstrike/entity/PlayerPawn.h"
#include "cstrike/interface/CGameEntitySystem.h"
#include "cstrike/interface/IRecipientFilter.h"
#include "cstrike/interface/ISoundOpSystem.h"
#include "cstrike/type/CRecipientFilter.h"
#include "cstrike/type/EmitSound.h"

#include <cstdarg>

CBasePlayerController* CBasePlayerController::FindBySlot(PlayerSlot_t slot)
{
    return g_pGameEntitySystem->FindEntityByIndex<CBasePlayerController*>(slot + 1);
}

CBasePlayerPawn* CBasePlayerController::GetPawnInternal()
{
    return g_pGameEntitySystem->FindEntityByEHandle<CBasePlayerPawn*>(m_hPawn());
}

SteamId_t CBasePlayerController::GetSteamID()
{
    constexpr SteamId_t UserIdOffset = 76561197960265728;
    static auto         offset       = GetFieldOffset("CBasePlayerController", "m_steamID");
    const auto          steamId      = GetFieldValue<SteamId_t>(offset);
    if (IsFakeClient() || steamId < UserIdOffset)
        return 0;

    return steamId;
}

bool CBasePlayerController::CanHearAndReadChatFrom(CBasePlayerController* pSenderController)
{
    DeclareVFuncIndex(CBasePlayerController, CanHearAndReadChatFrom, offset);
    return VCall_Manual(offset, bool, this, pSenderController);
}

void CBasePlayerController::SwitchTeam(CStrikeTeam_t team)
{
    address::server::CBasePlayerController_SwitchSteam(this, team);
}

void CBasePlayerController::SetClanTag(const char* tag)
{
    const auto& clan = g_pGameEntitySystem->AllocPooledString(tag);
    m_szClan(clan);
}

void CBasePlayerController::ClientPrint(HudPrint_t dest, const char* name, const char* param1, const char* param2, const char* param3, const char* param4)
{
    if (IsBot() || !IsConnected())
        return;

    UTIL_TextMsgSingle(GetPlayerSlot(), dest, name, param1, param2, param3, param4);
}

void CBasePlayerController::ClientPrintF(HudPrint_t dest, const char* message, ...)
{
    va_list args;
    va_start(args, message);

    char buffer[256];
    vsnprintf(buffer, sizeof(buffer), message, args);
    va_end(args);

    ClientPrint(dest, buffer);
}

SndOpEventGuid_t CBasePlayerController::EmitSoundClient(const char* pszSound, const float* pVolume) const
{
    CSingleRecipientFilter filter(GetPlayerSlot());
    const auto             guid = g_pSoundOpGameSystem->StartSoundEvent(&filter, pszSound);

    if (pVolume != nullptr && guid.IsValid())
    {
        const auto         volume = *pVolume;
        CSosFieldDataFloat param(volume);
        g_pSoundOpGameSystem->SetSoundEventParam(&filter, guid, SE_PARAM_NAME_VOLUME, &param);
    }

    return guid;
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

CCSPlayerPawn* CCSPlayerController::GetPlayerPawn()
{
    return g_pGameEntitySystem->FindEntityByEHandle<CCSPlayerPawn*>(m_hPlayerPawn());
}

CCSObserverPawn* CCSPlayerController::GetObserverPawn()
{
    return g_pGameEntitySystem->FindEntityByEHandle<CCSObserverPawn*>(m_hObserverPawn());
}

void CCSPlayerController::ForceChangeTeam(CStrikeTeam_t team)

{
    address::server::CBasePlayerController_CheckPawn(this);
    ChangeTeam(team);
}

void CCSPlayerController::CheckPawn()
{
    address::server::CBasePlayerController_CheckPawn(this);
}

void CCSPlayerController::SetPawn(CCSPlayerPawn* pawn, bool unk1, bool unk2, bool unk3, bool unk4)
{
    address::server::CBasePlayerController_SetPawn(this, pawn, unk1, unk2, unk3, unk4);
}

void CCSPlayerController::RoundRespawn()
{
    // 在CCSGameRules::RestartRound的复活中
    // 有一个Func是跟VPhysics有关的
    // 暂时不知道是干什么的

    if (this->m_hPlayerPawn().ToInt() != this->m_hPawn().ToInt())
    {
        if (const auto pawn = g_pGameEntitySystem->FindEntityByEHandle<CCSPlayerPawn*>(this->m_hPlayerPawn()))
            SetPawn(pawn, true, false, false, false);
    }

    static auto offset = g_pGameData->GetVFunctionIndex("CBasePlayerController::RoundRespawn");
    CALL_VIRTUAL(void, offset, this);
}

CEconItemView* CCSPlayerController::GetItemInLoadoutFromInventory(CStrikeTeam_t team, int32_t slot)
{
    const auto pService = m_pInventoryServices();
    AssertPtr(pService);
    return pService->GetItemInLoadout(team, slot);
}